File: crates/paimon/src/table/read_builder.rs
// Licensed to the Apache Software Foundation (ASF) under one
// or more contributor license agreements.  See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership.  The ASF licenses this file
// to you under the Apache License, Version 2.0 (the
// "License"); you may not use this file except in compliance
// with the License.  You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing,
// software distributed under the License is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
// KIND, either express or implied.  See the License for the
// specific language governing permissions and limitations
// under the License.

//! ReadBuilder and TableRead for table read API.
//!
//! Reference: [pypaimon.read.read_builder.ReadBuilder](https://github.com/apache/paimon/blob/master/paimon-python/pypaimon/read/read_builder.py)
//! and [pypaimon.table.file_store_table.FileStoreTable.new_read_builder](https://github.com/apache/paimon/blob/master/paimon-python/pypaimon/table/file_store_table.py).

use super::{ArrowRecordBatchStream, Table, TableScan};
use crate::arrow::ArrowReaderBuilder;
use crate::spec::DataField;
use crate::DataSplit;

/// Builder for table scan and table read (with_projection, new_scan, new_read).
///
/// Reference: [pypaimon.read.read_builder.ReadBuilder](https://github.com/apache/paimon/blob/master/paimon-python/pypaimon/read/read_builder.py)
#[derive(Debug, Clone)]
pub struct ReadBuilder<'a> {
    table: &'a Table,
    projection: Option<Vec<String>>,
}

impl<'a> ReadBuilder<'a> {
    pub(crate) fn new(table: &'a Table) -> Self {
        Self {
            table,
            projection: None,
        }
    }

    /// Set projection (column names to read). If not set, all columns are read.
    pub fn with_projection(mut self, projection: Vec<String>) -> Self {
        self.projection = Some(projection);
        self
    }

    /// Create a table scan. Call [TableScan::plan] to get splits.
    pub fn new_scan(self) -> TableScan<'a> {
        TableScan::new(self.table)
    }

    /// Create a table read for consuming splits (e.g. from a scan plan).
    pub fn new_read(self) -> TableRead<'a> {
        let read_type = self.read_type();
        TableRead {
            table: self.table,
            read_type,
        }
    }

    /// Fields to read (projected or full schema). Used by [TableRead] and predicate builder.
    fn read_type(&self) -> Vec<DataField> {
        let fields = self.table.schema().fields();
        match &self.projection {
            None => fields.to_vec(),
            Some(proj) => {
                let name_to_field: std::collections::HashMap<_, _> = fields
                    .iter()
                    .map(|f| (f.name().to_string(), f.clone()))
                    .collect();
                proj.iter()
                    .filter_map(|name| name_to_field.get(name).cloned())
                    .collect()
            }
        }
    }
}

/// Table read: reads data from splits (e.g. produced by [TableScan::plan]).
///
/// Reference: [pypaimon.read.table_read.TableRead](https://github.com/apache/paimon/blob/master/paimon-python/pypaimon/read/table_read.py)
#[derive(Debug, Clone)]
pub struct TableRead<'a> {
    table: &'a Table,
    read_type: Vec<DataField>,
}

impl<'a> TableRead<'a> {
    /// Schema (fields) that this read will produce.
    pub fn read_type(&self) -> &[DataField] {
        &self.read_type
    }

    /// Table for this read.
    pub fn table(&self) -> &Table {
        self.table
    }

    pub fn to_arrow(&self, data_splits: &[DataSplit]) -> crate::Result<ArrowRecordBatchStream> {
        let arrow_reader_builder = ArrowReaderBuilder::new(self.table.file_io.clone()).build();
        arrow_reader_builder.read(data_splits)
    }
}

File: bindings/go/read_builder.go
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

package paimon

import (
	"context"
	"runtime"
	"sync"
	"unsafe"

	"github.com/jupiterrider/ffi"
)

// ReadBuilder creates TableScan and TableRead instances.
type ReadBuilder struct {
	ctx       context.Context
	lib       *libRef
	inner     *paimonReadBuilder
	closeOnce sync.Once
}

// Close releases the read builder resources. Safe to call multiple times.
func (rb *ReadBuilder) Close() {
	rb.closeOnce.Do(func() {
		ffiReadBuilderFree.symbol(rb.ctx)(rb.inner)
		rb.inner = nil
		rb.lib.release()
	})
}

// WithProjection sets column projection by name. Output order follows the
// caller-specified order. Unknown or duplicate names cause NewRead() to fail;
// an empty list is a valid zero-column projection.
func (rb *ReadBuilder) WithProjection(columns []string) error {
	if rb.inner == nil {
		return ErrClosed
	}
	projFn := ffiReadBuilderWithProjection.symbol(rb.ctx)
	return projFn(rb.inner, columns)
}

// NewScan creates a TableScan for planning which data files to read.
func (rb *ReadBuilder) NewScan() (*TableScan, error) {
	if rb.inner == nil {
		return nil, ErrClosed
	}
	createFn := ffiReadBuilderNewScan.symbol(rb.ctx)
	inner, err := createFn(rb.inner)
	if err != nil {
		return nil, err
	}
	rb.lib.acquire()
	return &TableScan{ctx: rb.ctx, lib: rb.lib, inner: inner}, nil
}

// NewRead creates a TableRead for reading data from splits.
func (rb *ReadBuilder) NewRead() (*TableRead, error) {
	if rb.inner == nil {
		return nil, ErrClosed
	}
	createFn := ffiReadBuilderNewRead.symbol(rb.ctx)
	inner, err := createFn(rb.inner)
	if err != nil {
		return nil, err
	}
	rb.lib.acquire()
	return &TableRead{ctx: rb.ctx, lib: rb.lib, inner: inner}, nil
}

var ffiReadBuilderFree = newFFI(ffiOpts{
	sym:    "paimon_read_builder_free",
	rType:  &ffi.TypeVoid,
	aTypes: []*ffi.Type{&ffi.TypePointer},
}, func(_ context.Context, ffiCall ffiCall) func(rb *paimonReadBuilder) {
	return func(rb *paimonReadBuilder) {
		ffiCall(
			nil,
			unsafe.Pointer(&rb),
		)
	}
})

var ffiReadBuilderWithProjection = newFFI(ffiOpts{
	sym:    "paimon_read_builder_with_projection",
	rType:  &ffi.TypePointer,
	aTypes: []*ffi.Type{&ffi.TypePointer, &ffi.TypePointer},
}, func(ctx context.Context, ffiCall ffiCall) func(rb *paimonReadBuilder, columns []string) error {
	return func(rb *paimonReadBuilder, columns []string) error {
		var colPtrs []*byte
		var cStrings [][]byte

		// Convert Go strings to null-terminated C strings
		for _, col := range columns {
			cStr := append([]byte(col), 0)
			cStrings = append(cStrings, cStr)
			colPtrs = append(colPtrs, &cStr[0])
		}
		// Null-terminate the array
		colPtrs = append(colPtrs, nil)

		var colsPtr unsafe.Pointer
		if len(colPtrs) > 0 {
			colsPtr = unsafe.Pointer(&colPtrs[0])
		}

		var errPtr *paimonError
		ffiCall(
			unsafe.Pointer(&errPtr),
			unsafe.Pointer(&rb),
			unsafe.Pointer(&colsPtr),
		)
		// Ensure Go-managed buffers stay alive for the full native call.
		runtime.KeepAlive(cStrings)
		runtime.KeepAlive(colPtrs)
		if errPtr != nil {
			return parseError(ctx, errPtr)
		}
		return nil
	}
})

var ffiReadBuilderNewScan = newFFI(ffiOpts{
	sym:    "paimon_read_builder_new_scan",
	rType:  &typeResultTableScan,
	aTypes: []*ffi.Type{&ffi.TypePointer},
}, func(ctx context.Context, ffiCall ffiCall) func(rb *paimonReadBuilder) (*paimonTableScan, error) {
	return func(rb *paimonReadBuilder) (*paimonTableScan, error) {
		var result resultTableScan
		ffiCall(
			unsafe.Pointer(&result),
			unsafe.Pointer(&rb),
		)
		if result.error != nil {
			return nil, parseError(ctx, result.error)
		}
		return result.scan, nil
	}
})

var ffiReadBuilderNewRead = newFFI(ffiOpts{
	sym:    "paimon_read_builder_new_read",
	rType:  &typeResultNewRead,
	aTypes: []*ffi.Type{&ffi.TypePointer},
}, func(ctx context.Context, ffiCall ffiCall) func(rb *paimonReadBuilder) (*paimonTableRead, error) {
	return func(rb *paimonReadBuilder) (*paimonTableRead, error) {
		var result resultNewRead
		ffiCall(
			unsafe.Pointer(&result),
			unsafe.Pointer(&rb),
		)
		if result.error != nil {
			return nil, parseError(ctx, result.error)
		}
		return result.read, nil
	}
})
